File: examples/agent_client.ts
/**
 * Minimal zinc-cli contract client for AI agents (TypeScript).
 *
 * Usage:
 *   export ZINC_WALLET_PASSWORD='...'
 *   npx tsx examples/agent_client.ts
 */

import { spawnSync } from "node:child_process";

type Json = Record<string, unknown>;

class ZincCliError extends Error {
  envelope: Json;

  constructor(envelope: Json) {
    const error = (envelope.error ?? {}) as Json;
    super(
      `zinc-cli error type=${String(error.type)} exit_code=${String(
        error.exit_code
      )}: ${String(error.message)}`
    );
    this.envelope = envelope;
  }
}

function runZinc(args: string[], correlationId?: string): Json {
  const cmd = ["--agent", ...(correlationId ? ["--correlation-id", correlationId] : []), ...args];
  const result = spawnSync("zinc-cli", cmd, { encoding: "utf-8" });
  const stdout = (result.stdout ?? "").trim();

  if (!stdout) {
    throw new Error(`zinc-cli returned empty stdout (stderr=${(result.stderr ?? "").trim()})`);
  }

  const envelope = JSON.parse(stdout) as Json;
  if (envelope.ok !== true) {
    throw new ZincCliError(envelope);
  }
  return envelope;
}

function main(): void {
  const cid = "ts-agent-demo-001";
  const info = runZinc(["wallet", "info"], cid);
  console.log(
    `profile=${String(info.profile)} network=${String(info.network)} scheme=${String(info.scheme)}`
  );

  runZinc(["sync", "chain"], cid);
  runZinc(["sync", "ordinals"], cid);
  const balance = runZinc(["balance"], cid);
  const total = ((balance.total ?? {}) as Json).confirmed;
  console.log(`confirmed_sats=${String(total)}`);
}

main();

File: examples/agent_client.py
#!/usr/bin/env python3
"""
Minimal zinc-cli contract client for AI agents.

Usage:
  export ZINC_WALLET_PASSWORD='...'
  python3 examples/agent_client.py
"""

from __future__ import annotations

import json
import subprocess
from typing import Any, Dict, List


class ZincCliError(RuntimeError):
    def __init__(self, envelope: Dict[str, Any]):
        self.envelope = envelope
        error = envelope.get("error", {})
        super().__init__(
            f"zinc-cli error type={error.get('type')} exit_code={error.get('exit_code')}: {error.get('message')}"
        )


def run_zinc(args: List[str], *, correlation_id: str | None = None) -> Dict[str, Any]:
    cmd = ["zinc-cli", "--agent"]
    if correlation_id:
        cmd += ["--correlation-id", correlation_id]
    cmd += args

    proc = subprocess.run(cmd, capture_output=True, text=True, check=False)
    stdout = proc.stdout.strip()
    if not stdout:
        raise RuntimeError(f"zinc-cli returned empty stdout (stderr={proc.stderr.strip()})")

    envelope = json.loads(stdout)
    if not envelope.get("ok", False):
        raise ZincCliError(envelope)
    return envelope


def main() -> None:
    cid = "py-agent-demo-001"
    info = run_zinc(["wallet", "info"], correlation_id=cid)
    print(f"profile={info.get('profile')} network={info.get('network')} scheme={info.get('scheme')}")

    # Read-only snapshot flow
    run_zinc(["sync", "chain"], correlation_id=cid)
    run_zinc(["sync", "ordinals"], correlation_id=cid)
    balance = run_zinc(["balance"], correlation_id=cid)
    confirmed = balance.get("total", {}).get("confirmed")
    print(f"confirmed_sats={confirmed}")


if __name__ == "__main__":
    main()
